### ISSUE TITLE
hirni: test_demo_repro_analysis failure in recent PRs

### ISSUE THREAD

**Issue by kyleam** [datalad/datalad#5549]
`datalad_hirni.tests.test_demos.test_demo_repro_analysis` has failed in two unrelated PRs (gh-5534, gh-5539).  It's reliably showing up in gh-5534, enough that I thought it must be a non-obvious interaction with those changes, but then it popped up in another PR, gh-5539.  I can't trigger it locally, and it's not consistently happening in PRs (aside from maybe gh-5534).

Any ideas what could be going here, @bpoldrack?

<details>
<summary>error</summary>

```
datalad_hirni.tests.test_demos.test_demo_repro_analysis ... /tmp/datalad_temp_test_demo_repro_analysiswoay53v2/sourcedata/code/hirni-toolbox/procedures/copy-converter.sh: line 9: $1: unbound variable
ERROR
[...]
======================================================================
ERROR: datalad_hirni.tests.test_demos.test_demo_repro_analysis
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/tests/utils.py", line 757, in _wrap_with_tempfile
    return t(*(arg + (filename,)), **kw)
. (dataset): 
acq100 (dataset): dicom
Versions: annexremote=1.5.0 appdirs=1.4.4 boto=2.49.0 cmd:7z=16.02 cmd:annex=8.20210223-1~ndall+1 cmd:bundled-git=2.24.0 cmd:git=2.24.0 cmd:system-git=2.31.0 cmd:system-ssh=8.2p1 exifread=2.3.2 humanize=3.3.0 iso8601=0.1.14 keyring=23.0.1 keyrings.alt=UNKNOWN msgpack=1.0.2 mutagen=1.45.1 requests=2.25.1 wrapt=1.12.1
Obscure filename: str=b' |;&%b5{}\'"\xce\x94\xd0\x99\xd7\xa7\xd9\x85\xe0\xb9\x97\xe3\x81\x82 .datc ' repr=' |;&%b5{}\'"ΔЙקم๗あ .datc '
Encodings: default='utf-8' filesystem='utf-8' locale.prefered='UTF-8'
Environment: LANG='C.UTF-8' PATH='/opt/hostedtoolcache/Python/3.7.10/x64/bin:/opt/hostedtoolcache/Python/3.7.10/x64:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/opt/pipx_bin:/usr/share/rust/.cargo/bin:/home/runner/.config/composer/vendor/bin:/usr/local/.ghcup/bin:/home/runner/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin' GIT_CONFIG_PARAMETERS="'init.defaultBranch=master'"
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/tests/utils.py", line 757, in _wrap_with_tempfile
    return t(*(arg + (filename,)), **kw)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad_hirni/tests/utils.py", line 257, in newfunc
    return f(*(arg + (new_url,)), **kw)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad_hirni/tests/test_demos.py", line 156, in test_demo_repro_analysis
    anonymize=True)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/distribution/dataset.py", line 505, in apply_func
    return f(**kwargs)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/interface/utils.py", line 482, in eval_func
    return return_func(generator_func)(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/interface/utils.py", line 470, in return_func
    results = list(results)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/interface/utils.py", line 401, in generator_func
    allkwargs):
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/interface/utils.py", line 557, in _process_results
    for res in results:
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad_hirni/commands/spec2bids.py", line 268, in __call__
    return_type='generator'
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/interface/utils.py", line 401, in generator_func
    allkwargs):
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/interface/utils.py", line 557, in _process_results
    for res in results:
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/interface/run_procedure.py", line 451, in __call__
    return_type='generator'
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/interface/utils.py", line 401, in generator_func
    allkwargs):
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/interface/utils.py", line 557, in _process_results
    for res in results:
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/core/local/run.py", line 255, in __call__
    sidecar=sidecar):
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/core/local/run.py", line 715, in run_command
    raise exc
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/core/local/run.py", line 490, in _execute_command
    command
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/cmd.py", line 412, in run
    **results,
datalad.support.exceptions.CommandError: CommandError: 'bash /tmp/datalad_temp_test_demo_repro_analysiswoay53v2/sourcedata/code/hirni-toolbox/procedures/copy-converter.sh ' failed with exitcode 1 under /tmp/datalad_temp_test_demo_repro_analysiswoay53v2
```

</details>

https://github.com/datalad/datalad/pull/5534#issuecomment-811459153
https://github.com/datalad/datalad/runs/2240654101
https://github.com/datalad/datalad/runs/2254534141

**Comment by bpoldrack** [comment]
Not obvious, no. What fails is this procedure:

```
#!/usr/bin/env bash
# simply copy a file and add it to destination dataset
# TODO: a bit of protection for too many/less arguments ...

set -e -u
set -x

if [ -d "$1" ]
then
    datalad run --explicit --input "$1" --output "$2"  cp -r "$1" "$2"
else
    datalad run --explicit --input "$1" --output "$2"  cp "$1" "$2"
fi
```

I don't see anything wrong with the run calls, so I guess something is screwing the arguments it is called with.
Investigating ...

**Comment by bpoldrack** [comment]
Ok. So: An explicit `reload` (no force) of the `ConfigManager` of the subdatasets inspected by `_get_procedure_implementation` seems to fix it. However, this raises two questions:

1. Famous state of debugging "How did it ever work?": Why would that issue not always be triggered? Why would it resolve itself after several calls to that function (see above)? Can't see any implicit reload anywhere.
2. Is it okay to let `run-procedure` call reload on all inspected subdatasets? Again, it's `force=False` so actually shouldn't be incredibly expensive, but it's not free either and will often(?) be superfluous.

WDYT, @datalad/developers ?

**Comment by bpoldrack** [comment]
To clarify (1):

If it only was the first part, an answer seems obvious: If there wasn't a `Dataset` and `*Repo` instance for a subdataset around already, we would get a fresh `ConfigManager` - no reload required. But that doesn't explain, how it manages to resolve itself after some time/some calls.

**Comment by yarikoptic** [comment]
> 1. Famous state of debugging "How did it ever work?": Why would that issue not always be triggered? Why would it resolve itself after several calls to that function (see above)? Can't see any implicit reload anywhere.

code is full of `.reload` calls (without force) at many places were we suspect situation with config might have been changed due to our or external actions.  I believe at some point I proposed more aggressive use of it within base GitRepo interactions with config (https://github.com/datalad/datalad/pull/4364) but we decided against it.  So it remains for us to spot such other places, and in general to come up with a clear formulation on our expectation for "life of the .config" (as I have mentioned in https://github.com/datalad/datalad/pull/4364#issuecomment-607288130).  So, I guess go with (2).

**Comment by bpoldrack** [comment]
My main concern actually is, that there seems to be at least one missing piece, if not two. What exactly triggers the situation (other hirni tests seem to not fail while doing "the same thing")? And how did it resolve itself, when there appears to be no reload call? If there actually is an implicit reload somewhere, may be that would be a better spot or should be removed in favor of the new reload.

**Comment by kyleam** [comment]
I'm still trying to trigger this locally in hopes of having something helpful to say about the remaining mysteries, but thanks very much, @bpoldrack, for the detailed analysis.

**Comment by kyleam** [comment]
> What exactly triggers the situation (other hirni tests seem to not fail while doing "the same thing")? And how did it resolve itself, when there appears to be no reload call? If there actually is an implicit reload somewhere, may be that would be a better spot or should be removed in favor of the new reload.

I've still had no luck triggering the `test_demo_repro_analysis` failure on my end, and I can't answer these questions.  However, based on your explanation and on the minimal test I added in gh-5552, it seems to me that the most likely reason for the failure _not_ being triggered is that `_get_procedure_implementation` is the first to create dataset instance at path.  When the test passes on my end, that is indeed what seems to trigger the `reload` call for hirni-toolbox.

But, again, that doesn't explain what's aggravated the test in gh-5534 and, to a lesser degree, in gh-5539 (presumably by instantiating the dataset before `_get_procedure_implementation` does), or how it resolves itself.

<details>
<summary>details</summary>

Adding these debugging statements...

```diff
diff --git a/datalad_hirni/commands/spec2bids.py b/datalad_hirni/commands/spec2bids.py
index 2197e0e..793556c 100644
--- a/datalad_hirni/commands/spec2bids.py
+++ b/datalad_hirni/commands/spec2bids.py
@@ -263,6 +263,7 @@ def __call__(specfile, dataset=None, anonymize=False, only_type=None):
                         # apparently reload is necessary to consider config
                         # overrides via env:
                         dataset.config.reload()
+                        print("&&&&&&&&& Start of run_procedure")
                         for r in dataset.run_procedure(
                                 spec=proc_name,
                                 return_type='generator'
diff --git a/datalad_hirni/tests/test_demos.py b/datalad_hirni/tests/test_demos.py
index 36d0d7b..77782ac 100644
--- a/datalad_hirni/tests/test_demos.py
+++ b/datalad_hirni/tests/test_demos.py
@@ -150,11 +150,11 @@ def test_demo_repro_analysis(bids_path, ana_path, toolbox_url):
     assert_result_count(subs, 1, path=op.join(localizer_ds.path, 'sourcedata', 'code', 'hirni-toolbox'))
     assert_result_count(subs, 1, path=op.join(localizer_ds.path, 'sourcedata', 'acq1', 'dicoms'))
     assert_result_count(subs, 1, path=op.join(localizer_ds.path, 'sourcedata', 'acq2', 'dicoms'))
-
+    print(">>>>>>> Starting execution of failing line")
     localizer_ds.hirni_spec2bids([op.join(localizer_ds.path, 'sourcedata', 'studyspec.json')] +
                                  glob.glob(op.join(localizer_ds.path, 'sourcedata', '*', 'studyspec.json')),
                                  anonymize=True)
-
+    print(">>>>>>> Finished execution of failing line")
     for f in ['sub-001',
               'task-oneback_bold.json',
               'participants.tsv',
```

```diff
diff --git a/datalad/config.py b/datalad/config.py
index 6f57822d15..e9b7f8cceb 100644
--- a/datalad/config.py
+++ b/datalad/config.py
@@ -397,7 +397,11 @@ def reload(self, force=False):
         while to_run:
             store_id, runargs = to_run.popitem()
             self._stores[store_id] = self._reload(runargs)
-        print(f"reload called: {self._repo_pathobj}")
+        print(f"|||||||| reload called: {self._repo_pathobj}")
+        import traceback
+        import sys
+        traceback.print_stack(file=sys.stdout)
+        print(f"^^^^^^^^ end of trackback: {self._repo_pathobj}")
         # always update the merged representation, even if we did not reload
         # anything from a file. ENV or overrides could change independently
         # start with the commit dataset config
```

... here is the relevant part of the output:

```
|||||||| reload called: /tmp/datalad_temp_test_demo_repro_analysisemx58yn9/sourcedata/code/hirni-toolbox
  File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/kyle/src/python/venvs/dl-hirni/lib/python3.7/site-packages/nose/__main__.py", line 8, in <module>
    run_exit()
  File "/home/kyle/src/python/venvs/dl-hirni/lib/python3.7/site-packages/nose/core.py", line 121, in __init__
    **extra_args)
  File "/usr/lib/python3.7/unittest/main.py", line 101, in __init__
    self.runTests()
  File "/home/kyle/src/python/venvs/dl-hirni/lib/python3.7/site-packages/nose/core.py", line 207, in runTests
    result = self.testRunner.run(self.test)
  File "/home/kyle/src/python/venvs/dl-hirni/lib/python3.7/site-packages/nose/core.py", line 62, in run
    test(result)
  File "/home/kyle/src/python/venvs/dl-hirni/lib/python3.7/site-packages/nose/suite.py", line 178, in __call__
    return self.run(*arg, **kw)
  File "/home/kyle/src/python/venvs/dl-hirni/lib/python3.7/site-packages/nose/suite.py", line 225, in run
    test(orig)
  File "/home/kyle/src/python/venvs/dl-hirni/lib/python3.7/site-packages/nose/suite.py", line 178, in __call__
    return self.run(*arg, **kw)
  File "/home/kyle/src/python/venvs/dl-hirni/lib/python3.7/site-packages/nose/suite.py", line 225, in run
    test(orig)
  File "/home/kyle/src/python/venvs/dl-hirni/lib/python3.7/site-packages/nose/case.py", line 46, in __call__
    return self.run(*arg, **kwarg)
  File "/home/kyle/src/python/venvs/dl-hirni/lib/python3.7/site-packages/nose/case.py", line 134, in run
    self.runTest(result)
  File "/home/kyle/src/python/venvs/dl-hirni/lib/python3.7/site-packages/nose/case.py", line 152, in runTest
    test(result)
  File "/usr/lib/python3.7/unittest/case.py", line 663, in __call__
    return self.run(*args, **kwds)
  File "/usr/lib/python3.7/unittest/case.py", line 615, in run
    testMethod()
  File "/home/kyle/src/python/venvs/dl-hirni/lib/python3.7/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/kyle/src/python/datalad/datalad/tests/utils.py", line 757, in _wrap_with_tempfile
    return t(*(arg + (filename,)), **kw)
  File "/home/kyle/src/python/datalad/datalad/tests/utils.py", line 757, in _wrap_with_tempfile
    return t(*(arg + (filename,)), **kw)
  File "/home/kyle/src/python/datalad-hirni/datalad_hirni/tests/utils.py", line 257, in newfunc
    return f(*(arg + (new_url,)), **kw)
  File "/home/kyle/src/python/datalad-hirni/datalad_hirni/tests/test_demos.py", line 156, in test_demo_repro_analysis
    anonymize=True)
  File "/home/kyle/src/python/datalad/datalad/distribution/dataset.py", line 505, in apply_func
    return f(**kwargs)
  File "/home/kyle/src/python/datalad/datalad/interface/utils.py", line 482, in eval_func
    return return_func(generator_func)(*args, **kwargs)
  File "/home/kyle/src/python/datalad/datalad/interface/utils.py", line 470, in return_func
    results = list(results)
  File "/home/kyle/src/python/datalad/datalad/interface/utils.py", line 401, in generator_func
    allkwargs):
  File "/home/kyle/src/python/datalad/datalad/interface/utils.py", line 557, in _process_results
    for res in results:
  File "/home/kyle/src/python/datalad-hirni/datalad_hirni/commands/spec2bids.py", line 269, in __call__
    return_type='generator'
  File "/home/kyle/src/python/datalad/datalad/interface/utils.py", line 401, in generator_func
    allkwargs):
  File "/home/kyle/src/python/datalad/datalad/interface/utils.py", line 557, in _process_results
    for res in results:
  File "/home/kyle/src/python/datalad/datalad/interface/run_procedure.py", line 383, in __call__
    next(_get_procedure_implementation(name, ds=ds))
  File "/home/kyle/src/python/datalad/datalad/interface/run_procedure.py", line 138, in _get_procedure_implementation
    for m, n, f, h in _get_procedure_implementation(name=name, ds=subds):
  File "/home/kyle/src/python/datalad/datalad/interface/run_procedure.py", line 138, in _get_procedure_implementation
    for m, n, f, h in _get_procedure_implementation(name=name, ds=subds):
  File "/home/kyle/src/python/datalad/datalad/interface/run_procedure.py", line 127, in _get_procedure_implementation
    if ds is not None and ds.is_installed():
  File "/home/kyle/src/python/datalad/datalad/distribution/dataset.py", line 386, in is_installed
    self.repo is not None
  File "/home/kyle/src/python/datalad/datalad/distribution/dataset.py", line 291, in repo
    self._repo = repo_from_path(self._path)
  File "/home/kyle/src/python/datalad/datalad/core/local/repo.py", line 61, in repo_from_path
    repo = cls(path, create=False, **kw)
  File "/home/kyle/src/python/datalad/datalad/dataset/repo.py", line 154, in __call__
    instance = type.__call__(cls, *new_args, **new_kwargs)
  File "/home/kyle/src/python/datalad/datalad/support/annexrepo.py", line 240, in __init__
    config = self.config
  File "/home/kyle/src/python/datalad/datalad/support/gitrepo.py", line 1122, in config
    return self.cfg
  File "/home/kyle/src/python/datalad/datalad/dataset/gitrepo.py", line 169, in cfg
    self._cfg = ConfigManager(dataset=self, source='any')
  File "/home/kyle/src/python/datalad/datalad/config.py", line 346, in __init__
    self.reload(force=True)
  File "/home/kyle/src/python/datalad/datalad/config.py", line 403, in reload
    traceback.print_stack(file=sys.stdout)
^^^^^^^^ end of trackback: /tmp/datalad_temp_test_demo_repro_analysisemx58yn9/sourcedata/code/hirni-toolbox
```


So, `_get_procedure_implementation` triggers the reload through `Config.__init__`.  The environment variable is there at this point, so all is good.

</details>

**Comment by bpoldrack** [comment]
> I've still had no luck triggering the test_demo_repro_analysis failure on my end,

Oh well, it stopped reproducing for me, too! Same setup as yesterday but a reboot in between. I don't know. What I found and fixed seems to be a real issue, but it's not the whole picture.

**Comment by bpoldrack** [comment]
So, one additional piece: Trying to investigate the self-resolution of the issue mentioned above. This is: If the issue gets triggered, I drop into debugger within `run-procedure` right after its call to internal `_get_procedure_implementation`, which just yielded the wrong result. From debugger I execute `[x for x in _get_procedure_implementation(name, ds)]`. 5th or 6th subsequent execution then yields the correct result. To find out, how this happens I added a log message in `ConfigManager.reload()` to see whether it was called. And indeed, the execution of said command triggers additional reload calls when it does deliver the correct result, that are missing when it doesn't. However, I can't see yet where those reloads are coming from and therefore when and why exactly the do and do not happen.
This is because ATM I'm not able to add any information to that log message in `reload`. If I add at least the dataset path, the issue doesn't get triggered in the first place, if I enable log traceback - the issue doesn't get triggered ...
Looks like it's a `ds.is_installed` call, though, suggesting that its call to `Dataset.repo` plays a role.

I'll keep digging. But seems to me that I am right on some threshold of runtime that would trigger/not trigger it. Some time based effect - don't see how any of this would otherwise make sense.

**Comment by bpoldrack** [comment]
So, I think I got it. Once again - garbage collection!
An instance of the dataset (actually its `AnnexRepo`) in question (hirni-toolbox subdataset) exists, but isn't referenced anymore. Now, if the overall runtime hits the gc cycle that would pick up that instance before `run-procedure`'s internal call to `_get_procedure_implementation`, a new instance is created and `reload` is called - issue doesn't get triggered. If, however, gc didn't pick it up yet, we need an explicit `reload`. Same effect re self-resolution: There is no reference to that `AnnexRepo` being hold, so as soon as gc collects it, the same thing happens: We get a new instance and thereby an up-to-date `ConfigManager`.

Confirmed by printing the content of `AnnexRepo._unique_instances`. The one in question is present when the issue is triggered and absent when it's not.


Conclusion for me: The fix *is* a fix for it. It's just that the need for it didn't show earlier. There's no particular piece of code that as a side-effect triggers the issue, but we are simply at the edge of overall runtime in this case, where any minimal change can cause this call to be before or after the decisive gc cycle.

Agree, @datalad/developers ?
